### ISSUE TITLE
pid1: Slice= dbus property when set doesn#t follow rule that last assignment wins

### ISSUE THREAD

**Issue by intelfx** [systemd/systemd#20272]
**systemd version the issue has been seen with**
 > ```
 > systemd 249 (249-3-arch)
 > +PAM +AUDIT -SELINUX -APPARMOR -IMA +SMACK +SECCOMP +GCRYPT +GNUTLS +OPENSSL +ACL +BLKID +CURL +ELFUTILS +FIDO2 +IDN2 -IDN +IPTC +KMOD +LIBCRYPTSETUP +LIBFDISK +PCRE2 -PWQUALITY +P11KIT -QRENCODE +BZIP2 +LZ4 +XZ +ZLIB +ZSTD +XKBCOMMON +UTMP -SYSVINIT default-hierarchy=unified
 > ```

**Used distribution**
 > Arch Linux

**Linux kernel version used** (`uname -a`)

 > `Linux anystation 5.13.1-arch1pf3-1 #3 SMP PREEMPT Tue, 13 Jul 2021 13:44:10 +0000 x86_64 GNU/Linux`

**CPU architecture issue was seen on**
 > x86_64

**Expected behaviour you didn't see**

When `org.freedesktop.machine1.Manager.CreateMachineWithNetwork` is called with `Slice=foobar.slice` in scope_properties argument, systemd-machined is supposed to create the transient scope unit in the correct slice.

**Unexpected behaviour you saw**

The transient slice is started in `machine.slice`.

Specifically, systemd-machined attempts to invoke `org.freedesktop.systemd1.Manager.CreateTransientUnit` with both `Slice=machine.slice` and `Slice=foobar.slice` in the property array.

**Steps to reproduce the problem**

1. create a slice, say, `vm.slice`
2. create a libvirt qemu VM with `/vm` as the resource partition
3. configure conflicting properties on machine.slice and on the VM (say, set `AllowedCPUs=1` on `machine.slice`, but pin the VM to CPU 0 in the libvirt domain XML)
4. attempt to start the VM
5. observe the error returned by libvirt when it fails to set the affinity of the vCPU thread

**Additional program output to the terminal or log subsystem illustrating the issue**

Fragment of bus monitor log illustrating the problematic exchange:

```
$ sudo busctl monitor org.freedesktop.machine1
Monitoring bus message stream.
‣ Type=method_call  Endian=l  Flags=0  Version=1 Cookie=15  Timestamp="Thu 2021-07-22 04:18:40.744668 UTC"
  Sender=:1.90  Destination=org.freedesktop.machine1  Path=/org/freedesktop/machine1  Interface=org.freedesktop.machine1.Manager  Member=CreateMachineWithNetwork
  UniqueName=:1.90
  MESSAGE "sayssusaia(sv)" {
          STRING "qemu-2-windows";
          ARRAY "y" {
                  BYTE 8;
                  BYTE 161;
                  BYTE 102;
                  BYTE 210;
                  BYTE 228;
                  BYTE 239;
                  BYTE 71;
                  BYTE 250;
                  BYTE 158;
                  BYTE 126;
                  BYTE 191;
                  BYTE 94;
                  BYTE 16;
                  BYTE 78;
                  BYTE 72;
                  BYTE 33;
          };
          STRING "libvirt-qemu";
          STRING "vm";
          UINT32 141723;
          STRING "";
          ARRAY "i" {
                  INT32 14;
          };
          ARRAY "(sv)" {
                  STRUCT "sv" {
                          STRING "Slice";
                          VARIANT "s" {
                                  STRING "priority.slice";
                          };
                  };
                  STRUCT "sv" {
                          STRING "After";
                          VARIANT "as" {
                                  ARRAY "s" {
                                          STRING "libvirtd.service";
                                  };
                          };
                  };
                  STRUCT "sv" {
                          STRING "Before";
                          VARIANT "as" {
                                  ARRAY "s" {
                                          STRING "virt-guest-shutdown.target";
                                  };
                          };
                  };
          };
  };

‣ Type=method_call  Endian=l  Flags=0  Version=1 Cookie=47  Timestamp="Thu 2021-07-22 04:18:40.744810 UTC"
  Sender=:1.77  Destination=org.freedesktop.systemd1  Path=/org/freedesktop/systemd1  Interface=org.freedesktop.systemd1.Manager  Member=StartTransientUnit
  UniqueName=:1.77
  MESSAGE "ssa(sv)a(sa(sv))" {
          STRING "machine-qemu\x2d2\x2dwindows.scope";
          STRING "fail";
          ARRAY "(sv)" {
                  STRUCT "sv" {
                          STRING "Slice";
                          VARIANT "s" {
                                  STRING "machine.slice";
                          };
                  };
                  STRUCT "sv" {
                          STRING "Description";
                          VARIANT "s" {
                                  STRING "Virtual Machine qemu-2-windows";
                          };
                  };
                  STRUCT "sv" {
                          STRING "PIDs";
                          VARIANT "au" {
                                  ARRAY "u" {
                                          UINT32 141723;
                                  };
                          };
                  };
                  STRUCT "sv" {
                          STRING "Delegate";
                          VARIANT "b" {
                                  BOOLEAN true;
                          };
                  };
                  STRUCT "sv" {
                          STRING "CollectMode";
                          VARIANT "s" {
                                  STRING "inactive-or-failed";
                          };
                  };
                  STRUCT "sv" {
                          STRING "AddRef";
                          VARIANT "b" {
                                  BOOLEAN true;
                          };
                  };
                  STRUCT "sv" {
                          STRING "TasksMax";
                          VARIANT "t" {
                                  UINT64 16384;
                          };
                  };
                  STRUCT "sv" {
                          STRING "Slice";
                          VARIANT "s" {
                                  STRING "priority.slice";
                          };
                  };
                  STRUCT "sv" {
                          STRING "After";
                          VARIANT "as" {
                                  ARRAY "s" {
                                          STRING "libvirtd.service";
                                  };
                          };
                  };
                  STRUCT "sv" {
                          STRING "Before";
                          VARIANT "as" {
                                  ARRAY "s" {
                                          STRING "virt-guest-shutdown.target";
                                  };
                          };
                  };
          };
          ARRAY "(sa(sv))" {
          };
  };
```

**Comment by intelfx** [comment]
That said, I'm not sure who exactly is to blame. Reading the dbus spec a bit it looks like that is allowed behavior, and the last assignment is supposed to "win", like when writing units by hand:

> StartTransientUnit() may be used to create and start a transient unit which will be released as soon as it is not running or referenced anymore or the system is rebooted. name is the unit name including its suffix and must be unique. mode is the same as in StartUnit(), **properties contains properties of the unit, specified like in SetUnitProperties().**

> SetUnitProperties() may be used to modify certain unit properties at runtime. Not all properties may be changed at runtime, but many resource management settings (primarily those listed in systemd.resource-control(5)) may. The changes are applied instantly and stored on disk for future boots, unless runtime is true, in which case the settings only apply until the next reboot. name is the name of the unit to modify. **properties are the settings to set, encoded as an array of property name and value pairs. Note that this is not a dictionary!** Also note that when setting array properties with this method usually results in appending to the pre-configured array. To reset the configured arrays, set the property to an empty array first and then append to it.

However, this doesn't happen.
```
$ systemctl cat machine-qemu\\x2d1\\x2dwindows.scope 
# /run/systemd/transient/machine-qemu\x2d1\x2dwindows.scope
# This is a transient unit file, created programmatically via the systemd API. Do not edit.
[Scope]
Slice=machine.slice

[Unit]
Description=Virtual Machine qemu-1-windows

[Scope]
Delegate=yes

[Unit]
CollectMode=inactive-or-failed

[Scope]
TasksMax=16384
Slice=priority.slice

[Unit]
After=libvirtd.service
Before=virt-guest-shutdown.target
```
```
$ systemctl show machine-qemu\\x2d1\\x2dwindows.scope -p Slice
Slice=machine.slice
```

**Comment by poettering** [comment]
Closely related to #20182

**Comment by poettering** [comment]
Also closed by #21221
